android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt

```diff
@@ -70,124 +77,211 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
-    // TODO: choose a vibration pattern we like, and unset DEFAULT_VIBRATE.
-    builder.setVibrate(vPattern)
+    val title = when (fcmMessage.recipient) {
+        is Recipient.Stream -> "#${fcmMessage.recipient.stream} > ${fcmMessage.recipient.topic}"
+        // TODO use proper title for GroupPM, we will need
+        // to have a way to get names of PM users here.
+        is Recipient.GroupPm -> context.resources.getQuantityString(
+            R.plurals.group_pm,
+            fcmMessage.recipient.pmUsers.size - 2,
+            fcmMessage.sender.fullName,
+            fcmMessage.recipient.pmUsers.size - 2
+        )
+        is Recipient.Pm -> fcmMessage.sender.fullName
+    }
+    val isGroupConversation = when (fcmMessage.recipient) {
+        is Recipient.Stream -> true
+        is Recipient.GroupPm -> true
+        is Recipient.Pm -> false
+    }
+    val groupKey = extractGroupKey(fcmMessage.identity)
+    val conversationKey = extractConversationKey(fcmMessage)
+    val notification = getActiveNotification(context, conversationKey)
 
-    builder.setDefaults(NotificationCompat.DEFAULT_VIBRATE or NotificationCompat.DEFAULT_LIGHTS)
+    val messagingStyle = notification?.let {
+        NotificationCompat.MessagingStyle.extractMessagingStyleFromNotification(it)
+    } ?: NotificationCompat.MessagingStyle(selfUser)
+    messagingStyle
+        .setConversationTitle(title)
+        .setGroupConversation(isGroupConversation)
+        .addMessage(fcmMessage.content, fcmMessage.timeMs, sender)
 
-    val dismissIntent = Intent(context, NotificationIntentService::class.java)
-    dismissIntent.action = ACTION_CLEAR
-    val piDismiss = PendingIntent.getService(context, 0, dismissIntent, 0)
-    val action = NotificationCompat.Action(android.R.drawable.ic_menu_close_clear_cancel, "Clear", piDismiss)
-    builder.addAction(action)
+    val messageCount = messagingStyle.messages.size
+
+    val builder = NotificationCompat.Builder(context, CHANNEL_ID)
+    builder.apply {
+        // This should agree with `BRAND_COLOR` in the JS code.
+        color = Color.rgb(100, 146, 254)
+        if (BuildConfig.DEBUG) {
+            setSmallIcon(R.mipmap.ic_launcher)
+        } else {
+            setSmallIcon(R.drawable.zulip_notification)
+        }
+        setAutoCancel(true)
+        setStyle(messagingStyle)
+        setGroup(groupKey)
+        setSound(getNotificationSoundUri())
+        setContentIntent(createViewPendingIntent(fcmMessage, context))
+        setNumber(messageCount)
+
+        val extraData = Bundle()
+        extraData.putInt("lastZulipMessageId", fcmMessage.zulipMessageId)
+        extras = extraData
+    }
+
+    val summaryNotification = createSummaryNotification(context, fcmMessage, groupKey)
+
+    NotificationManagerCompat.from(context).apply {
+        notify(groupKey, NOTIFICATION_ID, summaryNotification.build())
+        notify(conversationKey, NOTIFICATION_ID, builder.build())
```

Comment:
This could use a line or two of comment saying that we rely entirely on the tags to distinguish notifications, and so the "IDs" can be a constant. I.e. summarizing the reasoning at https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719157582 .